src/yaml/from_yaml.cr

```diff
@@ -290,6 +290,38 @@ module Time::EpochMillisConverter
   end
 end
 
+module YAML::ArrayConverter(Converter)
+  def self.from_yaml(ctx : YAML::ParseContext, node : YAML::Nodes::Node) : Array
+    unless node.is_a?(YAML::Nodes::Sequence)
+      node.raise "Expected sequence, not #{node.class}"
+    end
+
+    ary = Array(typeof(Converter.from_yaml(ctx, node))).new
+
+    node.each do |value|
+      ary << Converter.from_yaml(ctx, value)
+    end
+
+    ary
+  end
+end
+
+module YAML::HashValueConverter(Converter)
+  def self.from_yaml(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+    unless node.is_a?(YAML::Nodes::Mapping)
+      node.raise "Expected mapping, not #{node.class}"
+    end
+
+    hash = Hash(String, typeof(Converter.from_yaml(ctx, node))).new
```

Comment:
(unlike JSON) YAML allows keys of different types, not just `String`.

Reply:
JSON allows this too, on the crystal types side. This can be improved later with a different converter if you want to convert the keys too.

Reply:
@Sija, that is true, we can improve it. I do believe that it could take another PR for that. Considering Hash also allows union types for the Key, the converter would need to handle the cast possibilities of different types, or would we make it strict?

Reply:
Both `Hash` converters simply need two generic arguments, one for converting the key, the other for converting the value. Just like `Hash` has two generic arguments as well.

Reply:
And I'd rather have this included here to avoid breaking the API later.

Reply:
I mean, it's called `HashValueConverter`, not `HashConverter`.

Reply:
So, I played a bit with a possible `YAML::HashConverter`, but wind up with some points.

Different from the basic types `Int32, String, Time, etc...`, the `Time::EpochConverter, Time::EpochMillisConverter` and others custom converters are modules which expose the method `self.from_yaml` that is used from the macro [mapping here](https://github.com/crystal-lang/crystal/blob/master/src/yaml/mapping.cr#L147).

So, to make a Hash works with union types, for example, `(Time::EpochConverter | String)` would be better to refactor these custom converters to have similar API of the basic types.

For example:
```crystal
struct Time::EpochConverter
  def from_yaml(ctx : YAML::ParseContext, node : YAML::Nodes::Node) : Time
    unless node.is_a?(YAML::Nodes::Scalar)
      node.raise "Expected scalar, not #{node.class}"
    end

    Time.unix(node.value.to_i)
  end
end
```

So, with that, we would avoid the situation commented on the snippet below.

```crystal
module YAML::HashConverter(K, V)
  def self.from_yaml(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
    unless node.is_a?(YAML::Nodes::Mapping)
      node.raise "Expected mapping, not #{node.class}"
    end

    hash = Hash(K, V).new

    YAML::Schema::Core.each(node) do |key, value|
      # here on the V.from_yaml(ctx, value)
      # It would work for if the type is Time::EpochConverter.from_yaml(ctx, value)
      # but in case it is a String, would be required something like String.new(ctx, value) instead
      hash[K.new(ctx, key)] = V.from_yaml(ctx, value)
    end

    hash
  end
end
```

Does it make sense @straight-shoota? Not sure if I missed something, so feel free to point out if I took the wrong path.

Concluding, I am happy to continue evolving this PR until we agree it is in a good shape, or either get it merged and open another PR with the changes + a `YAML::HashConverter` for keys and values.

Reply:
`String.from_yaml` should have the same shape as a converter - this is how it works for `JSON`.

Please make an issue about this. Would you be okay leaving the `YAML` stuff off before that is refactored? What would others prefer?

Reply:
Yes, please remove the hash converter and we can merge.

Reply:
@RX14 done. I will start another one soon with tha yaml hash converter.